tlatools/org.lamport.tlatools/src/tlc2/util/BufferedRandomAccessFile.java

```diff
@@ -5,149 +5,175 @@
+           f.lo <= f.curr < f.lo + BuffSz
 
        V3. Any (possibly) unflushed characters are stored
            in "f.buff":
 
-           (forall i in [f.lo, f.curr):
-             c(f)[i] == f.buff[i - f.lo])
+           forall i in BufferedIndexes:
+             c(f)[i] == f.buff[i - f.lo]
 
        V4. For all characters not covered by V3, c(f) and
            disk(f) agree:
 
-           (forall i in [f.lo, len(f)):
-             i not in [f.lo, f.curr) => c(f)[i] == disk(f)[i])
-
-       V5. "f.dirty" is true iff the buffer contains bytes that
-           should be flushed to the file; by V3 and V4, only part
-           of the buffer can be dirty.
-
-           f.dirty ==
-           (exists i in [f.lo, f.curr):
-             c(f)[i] != f.buff[i - f.lo])
-
-       V6. this.maxHi == this.lo + this.buff.length
-       
-       Note that "f.buff" can be "null" in a valid file, since the
-       range of characters in V3 is empty when "f.lo == f.curr".
-       
-       A file is said to be *ready* if the buffer contains the
-       current position, i.e., when:
-       
-       R1. !f.closed && f.buff != null
-           && f.lo <= f.curr && f.curr < f.hi
-           
-       When a file is ready, reading or writing a single byte
-       can be performed by reading or writing the in-memory
-       buffer without performing a disk operation.
+           forall i in [0, f.length):
+             i not in BufferedIndexes =>
+               c(f)[i] == disk(f)[i]
+
+       V5. If the buffer contains bytes that should be flushed to the
+           file then "f.dirty" is true.  Note that "f.dirty" can be
+           true even when there are no bytes that must be flushed (for
+           instance, if the client wrote the same bytes that are
+           already present on disk).
+
+           (exists i in BufferedIndexes: disk(f)[i] != f.buff[i - f.lo])
+             => f.dirty
+
+       CAUTION: The RandomAccessFile API allows clients to seek past
+       the end of the file.  In such a state, reads always return
+       EOF and writes expand the file with arbitrary data to be large
+       enough to contain the written bytes.  Therefore, it is possible
+       for the buffer to be past the end of the file (f.lo > f.length).
     */
 
-    /** Open a new <code>BufferedRandomAccessFile</code> on
-        <code>file</code> in mode <code>mode</code>, which
-        should be "r" for reading only, or "rw" for reading
-        and writing. */
-    public BufferedRandomAccessFile(File file, String mode)
-    throws IOException {
-      super(file, mode);
-      this.init();
+    /**
+     * Open a new <code>BufferedRandomAccessFile</code> on the
+     * file named <code>name</code> in mode <code>mode</code>,
+     * which should be "r" for reading only, or "rw" for reading
+     * and writing.
+     */
+    public BufferedRandomAccessFile(File file, String mode) throws IOException {
+        super(file, mode);
+        synchronized (mu) {
```

Comment:
The static buffer reuse is interesting.  Just to confirm my understanding: While it would be simpler if each new instance just created its own buffer each time, this appears to be a performance optimization to allow reuse of buffers in new files after old files are closed.  Then, even though this class isn't thread-safe if the same file is modified by two BRAFs, these `synchronized` sections are key to it being thread safe when _distinct_ files all are open with _different_ BRAFs.  Let me know if I've misinterpreted anything!

Did you consider adding this as part of the specification, given a naive danger of reading previous and different file's data?  I'm not sure it's necessary or worth it, as it seems correct, but figured it was worth discussion.

Reply:
To be precise:

 - Each BRAF has exclusive access to its internal buffer (`this.buff`).
 - Buffers can be re-used across instances. `availBuffs` is a global "pool" of previously-allocated buffers.
 - The lock `mu` allows BRAFs in different threads to safely interact with the global pool. They do this once on construction (to obtain a buffer) and once on `close()` (to return a buffer to the pool).

The purpose of the shared buffer pool is not clear to me. I suspect it is intended to reduce GC pressure: we can allocate buffers without using `new` and the garbage collector never needs to reclaim them.

However, I _strongly_ suspect that the optimization is irrelevant:
 - I don't believe TLC allocates/closes many instances of `BufferedRandomAccessFile`
 - Even if it does, the JVM's modern GC is _very_ good and may outperform this manual approach
 - And even if I'm wrong about those points, the `open()` syscall probably dominates `BufferedRandomAccessFile` allocation performance anyway

We could look at removing the shared pool in a separate PR.

> Did you consider adding this as part of the specification, given a naive danger of reading previous and different file's data?

I did not. I think the buffer pool problem is orthogonal, and there's a lot of evidence that the current approach works. (I didn't change it.) If I am wrong and it really is great for performance, then we could consider factoring it out into a separate class and specifying its behavior separately.

Reply:
That all tracks!  I think it makes the most sense to leave it as is for now to separate concerns and keep the focus of this PR.  And it makes sense to leave out of the spec given that, like you said, it wasn't changed and its future is dubious.

For discussions sake, as to the possible future of this approach, I agree that its benefit seems suspect.  Also, if there was need to reuse available buffers, the cost of `arraycopy` of all buffers on expansion seems odd.  If we suddenly opened and then closed 1000 files, this is a lot of copying.  I guess it's only the pointers (right?), given it's an array of arrays, but it's still strange to me.

Even if there was a benefit over JVM GC, it seems like atomic interactions with a `Stack` would provide a much more natural interface and should perform better on release.